### PR TITLE
Allows `add_file` to support an image URL

### PR DESCRIPTION
**Pull request by Elinvynia**
This is very shitty, tips on improving appreciated

### CODE REVIEW

File: src/http/client.rs [hunk 1]
```diff
@@ -1430,6 +1431,14 @@ impl Http {
                     multipart = multipart
                         .file(file_num.to_string(), path)?;
                 },
+                AttachmentType::Image(url) => {
+                    let mut picture: Vec<u8> = vec![];
+                    let req = &mut self.client.get(url).send()?;
+                    let _ = std::io::copy(&mut *req, &mut picture);
+                    multipart = multipart
+                        .part(file_num.to_string(), Part::bytes(Cow::Borrowed(&picture[..]).into_owned())
+                            .file_name(format!("{}.webp", file_num)));
```
Comment (Flat) [inline src/http/client.rs]: This should use the actual file extension unless you are converting the image to webp format.

File: src/http/client.rs [hunk 2]
```diff
@@ -1430,6 +1431,14 @@ impl Http {
                     multipart = multipart
                         .file(file_num.to_string(), path)?;
                 },
+                AttachmentType::Image(url) => {
+                    let mut picture: Vec<u8> = vec![];
+                    let req = &mut self.client.get(url).send()?;
+                    let _ = std::io::copy(&mut *req, &mut picture);
```
Comment (Flat) [inline src/http/client.rs]: You should handle the result of `std::io::copy` here instead of ignoring it with a `let _` binding

**Review by Flat — CHANGES_REQUESTED** [review]

File: src/http/client.rs [hunk 3]
```diff
@@ -1430,6 +1431,26 @@ impl Http {
...
+                    let path_segments = match parsed_url.path_segments() {
+                        Some(p) => p,
+                        None => return Err(Error::Url(url.to_string())),
+                    };
+                    let filename = match path_segments.last() {
+                        Some(l) => l,
+                        None => return Err(Error::Url(url.to_string())),
+                    };
+                    let mut picture: Vec<u8> = vec![];
+                    let req = &mut self.client.get(url).send()?;
```
Comment (arqunis) [inline src/http/client.rs]:
```suggestion
                    let mut req = self.client.get(url).send()?;
```

File: src/http/client.rs [hunk 4]
```diff
@@ -1430,6 +1431,26 @@ impl Http {
...
+                    std::io::copy(&mut *req, &mut picture)?;
```
Comment (arqunis) [inline src/http/client.rs]:
```suggestion
                    std::io::copy(&mut req, &mut picture)?;
```

File: src/http/client.rs [hunk 5]
```diff
@@ -1430,6 +1431,26 @@ impl Http {
                     multipart = multipart
                         .file(file_num.to_string(), path)?;
                 },
+                AttachmentType::Image(url) => {
+                    let parsed_url = match Url::parse(url) {
```
Comment (arqunis) [inline src/http/client.rs]:
```suggestion
                    let filename = Url::parse(url)
                        .ok()
                        .and_then(|url| url.path_segments())
                        .and_then(|segments| segments.last())
                        .ok_or_else(|| Error::Url(url.to_string())?;
```

File: src/http/mod.rs [hunk 6]
```diff
@@ -210,7 +212,13 @@ impl<'a> From<(&'a [u8], &str)> for AttachmentType<'a> {
 }
 
 impl<'a> From<&'a str> for AttachmentType<'a> {
-    fn from(s: &'a str) -> AttachmentType<'_> { AttachmentType::Path(Path::new(s)) }
+    fn from(s: &'a str) -> AttachmentType<'_> {
```
Comment (arqunis) [inline src/http/mod.rs:217]:
```suggestion
    /// Constructs an `AttachmentType` from a string.
    /// This string may refer to the path of a file on disk, or the http url to an image on the internet.
    fn from(s: &'a str) -> AttachmentType<'_> {
```

**Pull request merged**